base/regex.jl

```diff
@@ -282,22 +279,20 @@ function matchall(re::Regex, str::String, overlap::Bool=false)
     matches
 end
 
-matchall(re::Regex, str::SubString, overlap::Bool=false) =
+matchall(re::Regex, str::AbstractString, overlap::Bool=false) =
     matchall(re, String(str), overlap)
 
-function search(str::Union{String,SubString}, re::Regex, idx::Integer)
-    if idx > nextind(str,endof(str))
-        throw(BoundsError())
-    end
+function search(str::Union{String,SubString{String}}, re::Regex, idx::Integer)
+    isvalid(str, idx) || throw(ArgumentError("invalid index $i"))
     opts = re.match_options
     compile(re)
     PCRE.exec(re.regex, str, idx-1, opts, re.match_data) ?
         ((Int(re.ovec[1])+1):prevind(str,Int(re.ovec[2])+1)) : (0:-1)
 end
-search(s::AbstractString, r::Regex, idx::Integer) = throw(ArgumentError(
-    "regex search is only available for the String type; use String(s) to convert"
-))
-search(s::AbstractString, r::Regex) = search(s,r,start(s))
+search(str::Union{String,SubString{String}}, re::Regex) =
+    str == "" ? (0:-1) : search(str, re, start(s))
```

Comment:
I don't really like this, as it means passing the index `1` is not equivalent to omitting the index. Why not add this check the the method which takes an index? Also, the current behavior had its logic, as noted at https://github.com/JuliaLang/julia/issues/24103.

Better move this to a separate PR and only keep uncontroversial changes in this one. It already contains many different things.

Reply:
I will remove this change from this PR and add it to #24103.
I will discuss my rationale in #24103 (I have not read though it yet - I am working though this PR to fix #24157 first).
But in short passing `1` is exactly *not equivalent* to omitting the index. The corner case is an empty string "". You can search for something in an empty string (and it should not throw an error if you do not specify an index), but for this string index `1` is invalid so specifying it should be an error.